Feat/buildings routes

karolwolek opened the pull request.

czaja307 commented on app/models/building.ts:
```diff
@@ -24,10 +24,10 @@ export default class Building extends BaseModel {
   declare campusId: number
 
-  @column()
+  @column({ columnName: 'address_line1' })
```
why did you do it?
not saying it is wrong, just asking to see idea behind it

karolwolek replied:
so basically lucid is changing model field names to camel case in default. So addressLine1 will transfer to address_line_1 and thus its no valid in current migrations implementation (in diagrams and migrations its address_line1) so i thought less invasive is to set it here for address_line1 but i can change it other way around

czaja307 replied:
makes sense, your version looks cleaner too
@Lennox2536 you might want to check this out, because you are using the same columns

czaja307 commented on start/routes.ts:
```diff
@@ -9,15 +9,30 @@
+
+router
+  .group(() => {
+    // fetches campuses specifically
```
i think that i'd be more RESTful if you used separate endpoint for campuses

czaja307 commented on app/controllers/buildings_controller.ts:
```diff
@@ -0,0 +1,87 @@
+import Campus from '#models/campus'
+import { HttpContext } from '@adonisjs/core/http'
+export default class BuildingsController {
+  async getAll({ response }: HttpContext) {
```
the function names are okay, but i'd change them to more adonis conventional
if you use `node ace make:controller [name] --api`, you will get the api template

the conventional names are
index - list of all records
store - create new record
show - get individual record
update - quite obvious :p
destroy - deletes record

what i'd suggest - create separate 2 separate controllers to avoid naming conflicts, and change function names to index and store (getByCampus is fine imo)

karolwolek replied:
gotcha, will do

czaja307 commented on start/routes.ts:
```diff
@@ -9,15 +9,30 @@
+router
+  .group(() => {
+    // fetches campuses specifically
+    router.get('/campuses/:id', [BuildingsController, 'getByCampus'])
```
i think u used the wrong method here

czaja307 commented on database/seeders/building_seeder.ts:
seeders are a good idea for testing, but what u need to know is that they are executed in the alphabetical order and the campus_seeder needs to go before the buildings one, because of the dependencies

czaja307 commented on app/controllers/buildings_controller.ts:
```diff
@@ -0,0 +1,87 @@
+        return response.status(400).json({ message: 'Building ID is required' })
+      }
+      const building = await Building.find(id)
+      if (building) {
```
when you are doing checks like this it's better to be specific (js logic can be wild sometimes) so i'd go with `building !== null` to explicitly show what you meant

czaja307 commented on app/controllers/buildings_controller.ts:
```diff
@@ -0,0 +1,87 @@
+      }
+      const building = await Building.find(id)
+      if (building) {
+        return response.status(200).json(building.serialize())
```
i'm quite sure that you don't need to run `serialize()` method, the adonis api setup should handle it for you

mini-bomba replied:
verified, true

mini-bomba commented on app/controllers/buildings_controller.ts:
```diff
@@ -0,0 +1,87 @@
+      const campusId = request.param('id')
+      if (!campusId) return response.status(400).json({ message: 'campus ID is required' })
+
+      const [campus, buildings] = await Promise.all([
+        Campus.find(campusId),
+        Building.query().where('campusId', campusId),
+      ])
```
Couldn't we do the same thing using the `buildings` relationship on the `Campus` object?
(oh yeah also i've tried it and I think the relationship isn't defined correctly cuz it doesnt work, i have no idea why)

mini-bomba commented on app/controllers/buildings_controller.ts:
```diff
@@ -0,0 +1,87 @@
+    }
+  }
+
+  async getCampusById({ request, response }: HttpContext) {
+    try {
+      const id = request.param('id')
+      if (!id) return response.status(400).json({ message: 'campus ID is required' })
+
+      const campus = await Campus.find(id)
+      await campus?.load('buildings')
+      if (!campus) return response.status(404).json({ message: 'Campus not found' })
+
+      return response.status(200).json(campus.serialize())
+    } catch (error) {
+      return response.status(500).json({ message: 'Failed to fetch campus by Id' })
+    }
+  }
+}
```
This function appears unused.

mini-bomba commented on start/routes.ts:
```diff
@@ -9,15 +9,30 @@
+  .prefix('api/v1/files')
+
+router
+  .group(() => {
+    // fetches campuses specifically
+    router.get('/campuses/:id', [BuildingsController, 'getByCampus'])
+    router.get('/campuses', [BuildingsController, 'getCampuses'])
+
+    // fetches buildings specifically
+    router.get('in-campus/:id', [BuildingsController, 'getByCampus'])
+    router.get('/:id', [BuildingsController, 'getById'])
+    router.get('/', [BuildingsController, 'getAll'])
   })
-  .prefix('api/v1')
+  .prefix('api/v1/buildings')
```
1. `/api/v1/buildings/in-campus/:id` is a duplicate of `/api/v1/buildings/campuses/:id`
2. `/api/v1/buildings/campuses` doesn't seem like the most logical path to get a listing of campuses. Maybe this should be exposed outside of the buildings group? (`/api/v1/campuses`)
   Or maybe it would make more sense to start with campuses? `/api/v1/campuses` for a list of campuses, `/api/v1/campuses/buildings` for a list of all buildings, `/api/v1/campuses/:id/buildings` for a list of buildings in a given campus?

mini-bomba reviewed (commented):
Seeing seeders getting implemented make me somewhat hopeful... that maybe, at some point, we'll be doing some unit tests...
But there are a few issues with the implementation at the moment:
1. The seeders are creating test data, but aren't constrained to run only in a development environment. You probably should constrain them to development and testing environments only. ([see the relevant docs](https://lucid.adonisjs.com/docs/seeders#environment-specific-seeders))
2. You've created two seeders, but the building seeder relies on data created by the campus seeder. I think that we should create seeders per schema and not per table, as that way we'll eliminate the dependencies between seeders. 
   Cross-seeder dependencies are an easy way to make mistakes, as ACE will run the files in alphabetical order. (just like migrations) If you don't make sure that the seeder you depend on ends up first when sorting, you'll end up with errors when trying to run them, as happened when I tried to run the seeders after creating the database from scratch.
   ![image](https://github.com/user-attachments/assets/8b24452a-3fa5-437a-bd27-be816e44b6a5)

I also have some questions regarding the controller behaviour in edge cases:
1. Do we really need to return 404 when the query returns no results, but the query itself was valid? (specifically when the database has no building data, or when the provided campus ID was valid, but had no assigned buildings)
   IMO an empty list is already self-explanatory, 404 should be reserved for cases when the ID that the client provided was invalid.
2. Wrapping the controller methods in a try-catch seems a bit unnecessary to me. Adonis is already set up to catch errors and return 500 with the error message. The only thing this accomplishes is adding a generic context message. It also removes the log messages when such an error is thrown, which might make debugging harder.
   While the default adonis errors are quite bloated in development mode, they get reduced to basically the same thing as you're doing here with try/catches in production mode.
3. The API appears to return `campusId` fields of buildings as a string. I suspect this is because the foreign key column was defined as a `bigInteger` in the migrations. (cuz regular javascript integer numbers can only be 32bit, you need the bigint object to use 64bit numbers in js)
4. Is there a better way to validate the types of query parameters, one that isn't just manual checks in if statements?

karolwolek commented:
1. separated controllers into 2 different. Changed the methods mistake in router
2. rewrote endpoints
3. combined seeders together

mini-bomba commented on app/controllers/buildings_controller.ts:
When I mentioned 404s and empty lists, I did not mean to swap the 404s and 400s. The codes returned were correct.

What I meant was that in cases where the user makes a request like `GET /api/v1/buildings/in-campus/2`, and a campus with that ID exists, but has no associated buildings, should we really be returning an error? I don't think it's necessarily a bad way to handle such an edge case, but since an empty array is also a valid, we might as well return an empty array with a 200 status code. The same goes for the list all buildings endpoint.

I think that returning an empty array makes it easier for the client to figure out what went wrong and display an appropriate message. The client might want to display an actual error message when the ID it requested was invalid, but all it might want to do if there are no associated buildings is just show some "No buildings" text.
If you return 404 for both cases, the client would likely need to parse the response message, which would break if someone changed the message. Also each endpoint might have a different message that means "valid ID, no associated objects".

And I'm not saying that you should change the status codes; they're already correct for their use case, that's not a good way to solve this issue.
I guess a good way to solve the error message parsing issue would be to include a `code` field in the error response, and have the codes be shared across all endpoints. We could have an `E_NO_RESULTS` code for the case where you would return an empty list, and `E_UNKNOWN_ID` for the case where the ID was invalid. (we probably should do these even if we decide to just return empty arrays in this case)

Anyway, not all of my review comments are meant to be "hey, fix this". Some of them (like this one) are meant to be an invitation into a discussion.

czaja307 replied:
What i'd propose
1. In the `index` functions go with `404`, because in normal scenario there should always be some buildings to display.
2. The same for the show - this seems to be pretty common usage of `404` to me.
3. For the `getByCampus` we can use the similar logic if we are sure that every campus should have at least one building. If that's not the case i'd personally go with `204`, which signalises that request was successful but there are no records.

what do you guys think about this?

mini-bomba replied:
Adding a 204 case would make using the API more complex. (remember that 204 can never carry a response body, and even if you try, some clients will just ignore it)
IMO just returning an empty list with a 200 code would be the simplest.

karolwolek commented:
I had in mind @mini-bomba question about validation and i am also trying to implement custom validators for buildings and campuses using VineJS - adonis own validation framework.

karolwolek commented:
I have added required validators for user input in query params and route parameters. 
* The show methods will first ensure that the id is present in id column on specific table and only then attempt to fetch all columns. If id is not present vine.js will attach suitable error log in response.
* I added functionality to paginate response of lists of buildings and made it optional so if there are no query string params for pagination the api will try to send all records.
* I am not sure if changing BigInteger on foreigh key is a good approach, but i changed it for testing other functionalities so let me know about other ways for handling this
* Campuses routes have optional param for loading relational data from buildings by adding includeBuildings boolean param
* Buildings routes have optional param for loading relational data from campuses by adding includeCampus boolean param

karolwolek commented on database/migrations/1732381079405_create_buildings_table.ts:
Not sure how to handle this

dawidlinek reviewed (commented):
Great job! I'm especially impressed with your engagement during the code review process. I have a suggestion: consider creating model hooks to handle recurring functionalities. These could include not only pagination but also filtering, sorting, and even relation loading.
For inspiration, you can refer to the https://github.com/Solvro/backend-cocktail-api/blob/main/app/controllers/ingredients_controller.ts. However, keep in mind that this can be optimized and done even better.
With such hooks in place, you can easily implement listing methods without relying on specific methods like getByCampus. This approach avoids the need to create a getByX method for every field.
Regarding the message field in your listing, it's a matter of preference. You can also  consider structuring the response as `{ message, ...buildings }` to keep it more concise or throw out the message field in successful requests all together.

czaja307 commented on app/controllers/buildings_controller.ts:
```diff
@@ -1,70 +1,83 @@
+    const buildingsQuery = Building.query()
+    const message = 'List of buildings'
+
+    if (includeCampus) {
```
in general i think that checking specifically `if (includeCampus !== undefined && includeCampus === true)` helps to avoid JS logic madness
![image](https://github.com/user-attachments/assets/d0972aca-74a8-434d-8cd2-1be4b27448be)

czaja307 replied:
@mini-bomba global config coming soon™

czaja307 commented on app/controllers/buildings_controller.ts:
```diff
@@ -1,70 +1,83 @@
+    }
+
+    if (page) {
+      const buildings = await buildingsQuery.paginate(page, limit || 10)
```
`||` works okay in this case, but in general when you are checking for undefined it is better to use `??` operator

karolwolek commented on app/controllers/buildings_controller.ts:
```diff
@@ -1,70 +1,83 @@
+  async index({ request, response }: HttpContext) {
+    const { page, limit, includeCampus } = await request.validateUsing(indexValidator)
+
+    const buildingsQuery = Building.query()
```
Correct me if i am wrong but the ``` Building.all() ``` promises  to return an array of ```Building``` model. 
The ```Building.query()``` returns a query builder instance, allowing you to chain methods to add additional query parameters.

czaja307 replied:
didn't see that, you both are correct

mini-bomba reviewed (approved):
Epic work! :fire: :shipit: 

I've just got a few last minor comments.

karolwolek commented on app/models/building.ts:
```diff
@@ -3,13 +3,31 @@ import { DateTime } from "luxon";
 
+import { typedModel } from "../decorators/typed_model.js";
 import { BuildingIcon } from "../enums/building_icon.js";
+import { preloadRelations } from "../scopes/preload_helper.js";
+import { handleSearchQuery } from "../scopes/search_helper.js";
+import { handleSortQuery } from "../scopes/sort_helper.js";
```
thats a good point, will do

karolwolek replied:
I can't find a way to work around eslit screaming about those 
```ts
static preloadRelations = preloadRelations(Building);

static handleSearchQuery = handleSearchQuery(Building);

static handleSortQuery = handleSortQuery(Building);
```
when i defined imports as subpath. I will leave it for now. If you'll have time take a look if you want

karolwolek commented on app/scopes/preload_helper.ts:
```diff
@@ -0,0 +1,79 @@
+    return query;
+  }
+
+  const relationDefinition = model.$relationsDefinitions.get(relation);
+  if (relationDefinition === undefined) {
+    return query;
+  }
```
> Is this `relationDefinition === undefined` a case where the string that someone put in was invalid/had an invalid relation? 

Yes, my first idea was to simply ignore it bc I thought any error status response is not needed. The lack of preloaded data
sent to user is self explanatory. Do you mean any particular way of logging it ?

mini-bomba replied:
I see that the controllers filter out only allowed relations from the request to pass down to the scope, so it's possible that someone might make a typo while creating those relation allowlists.
I was thinking that something as simple as a `console.log` (or another logging equivalent) with a message that an invalid relation was given would work.
This would mainly be useful when testing/debugging your own controllers locally.

karolwolek commented on app/scopes/search_helper.ts:
```diff
@@ -0,0 +1,293 @@
+        >
+      >,
+    ) => {
+      for (const param in params) {
+        const value: string | string[] | FromTo | undefined = params[param];
```
just tried it and typescript can't resolve value's type with current approach for params type in function parameters definition. I got unknown and linter gets angry about it in further functions

karolwolek commented on database/seeders/building_seeder.ts:
i forgot about this comment and those seeders and i aggree. It was only for my specific tests cases. I don't even think it's anyway important in the main code anymore.

karolwolek commented on app/scopes/search_helper.ts:
```diff
@@ -0,0 +1,293 @@
+      break;
+    }
+    default: {
+      throw new NotImplementedException(
+        `Unsupported column type '${columnType}' in arrayTypeCheckHelper.`,
+      );
```
You left me thinking about this one and actually i could just apply here the same logic as for arrays: simply pass array with one value.
The code would change to sth like this 
```
      function handleDirectValue(
  query: ModelQueryBuilderContract<LucidModel, LucidRow>,
  column: ModelColumnOptions,
  value: string,
) {
  const columnType = extractType(column);
  const invalid = arrayTypeCheckHelper([value], columnType);
  if(invalid.length > 0) {
    throw new BadRequestException(
      `invalid filter value '${value}' ` +
        `for column: '${column.columnName}' ` +
        `of type: '${columnType}'`,
    );
  } else {
    if(columnType === "string") {
      query = query.whereILike(column.columnName, value);
    } else {
      query = query.where(column.columnName, value)
    }
    return query;
  }
}
```

mini-bomba replied:
:fire::fire::fire:

karolwolek commented on app/controllers/buildings_controller.ts:
```diff
@@ -0,0 +1,57 @@
+    const { page, limit } = await request.validateUsing(paginationValidator);
+    const baseQuery = Building.query().withScopes((scopes) => {
+      scopes.handleSearchQuery(
+        request.only([
+          "id",
+          "identifier",
+          "specialName",
+          "campusId",
+          "addressLine1",
+          "addressLine2",
+          "latitude",
+          "longitude",
+          "haveFood",
+          "createdAt",
+          "updatedAt",
+        ]),
+      );
```
> Or maybe the scope could just take the whole request and extract what it needs by itself?

Upside for manual pointing columns is to exclude and include just the columns you need. I intentionally excluded iconType column here since it does not have supported filtering type. This behaviour could be achieved by filtering object columns that have this type defined in meta property. But than again we loose the flexibility to exclude some other odd columns.

> Is there a way to do this without hard-coding every possible field in the controller?
```suggestion
        request.only([
          ...Object.keys(
            Building.$columnsDefinitions.entries().filter(([_, column]) => {
              const columnDef: ColumnDef = column;
              return columnDef.meta !== undefined
                ? columnDef.meta.type !== undefined
                : false;
            }),
          ),
        ]),
```
Could be done like so, dunno if I should even try to exclude columns with type other than allowed in this scope. Is it better than hardcoding? Hardcoded is more readable for sure, But suggested approach could  be more managable since you can get caught relying on word that "yes there's type for sure"

karolwolek replied:
Looking at my suggestion i've got an idea that maybe it's actually a good approach to move this logic inside the scope as you said and leave an option to exclude some columns by user as a side parameters and only those would be hardcoded ?

mini-bomba replied:
that's actually a good point, I didn't even notice that the iconType field was missing from that list.
I think that it's more likely that we'd want to exclude just a few fields rather than including just a few select ones.

czaja307 commented on app/controllers/campuses_controller.ts:
```diff
@@ -0,0 +1,38 @@
+      scopes.handleSortQuery(request.input("sort"));
+    });
+
+    return campuses;
```
nitpick: you can return campuses inline, instead of creating the variable for it

czaja307 commented on app/decorators/typed_model.ts:
```diff
@@ -0,0 +1,81 @@
+      constructor.$columnsDefinitions;
+
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    if (columns === undefined) {
```
question(linter): @Rei-x can you take a look at this line?

mini-bomba replied:
My guess is that because that field in the LucidModel is defined as just a Map, with no possibility of an undefined value, eslint can see that `columns` will never actually be undefined, even though it's defined here as possibly undefined.
![image](https://github.com/user-attachments/assets/c8f9fd72-ff58-4055-ac46-10b7fdcfabfe)

czaja307 commented on app/exceptions/not_implemented_exception.ts:
praise: custom exceptions look really cool

czaja307 commented on app/scopes/preload_helper.ts:
```diff
@@ -0,0 +1,79 @@
+ * const buildings = await Building.query()
+ *  .withScopes((scopes) => {
+ *    scopes.preloadRelations(request.only(["campus"]));
+ */
```
```suggestion
 *  });
```

czaja307 commented on app/scopes/preload_helper.ts:
```diff
@@ -0,0 +1,79 @@
+
+/**
+ * Helper function to work with recursively looaded chained relations
+ * - Base case [1] of function is the exhaustion of relationParts array (stack)
```
```suggestion
 * Helper function to work with recursively loaded chained relations
```

czaja307 replied:
github messed it up, it's about typo in `looaded`

czaja307 reviewed (dismissed):
really good job 🔥
